### ISSUE TITLE
Enhancement: [prefer-nullish-coalescing] should support `if` statement assignment (`??=`)

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#10829]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/prefer-nullish-coalescing/

### Description

I propose that the rule should flag `if` statements that are equivalent to a nullish coalescing assignment (`??=`).

### Fail

```TypeScript
declare let foo: { a: string } | null;

declare function makeFoo(): { a: string };

function lazyInitializeByTruthiness() {
  if (!foo) {
    foo = makeFoo();
  }
}

function lazyInitializeByNullCheck() {
  if (foo == null) {
    foo = makeFoo();
  }
}

export { }
```

### Pass

```TypeScript
declare let foo: { a: string } | null;

declare function makeFoo(): { a: string };

function lazyInitializeByNullishCoalescingAssignment() {
  foo ??= makeFoo();
}

export { }
```

### Additional Info

[playground](https://typescript-eslint.io/play/#ts=5.7.2&fileType=.ts&code=CYUwxgNghgTiAEEQBd4DMD2GBc8De8UuAzsjAJYB2A5vAL7wA%2B8lArhBANwBQ3oksBGlaUwychkrwAtlADWIAGJYAFAEpcBIvFIUa9Ht2GjxkxFABeATwCSlcuKgRyFkACErAFRitkACyoQYmJ1fG54eHI0eBUAQkwMNTCIiIT4AF4ZeSVVNR4Ium5CoxExCSloazsHcicXdysAOXYIAGE-cDlQvHDI6JU09My2DiSelPQsDKyFZQx1fPoi3hAADwAHDBhUAjogA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WlqYSNkAC1pkA9gEMkyMswDm6KL2jjokcGAC%2BILUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

relates to https://github.com/typescript-eslint/typescript-eslint/issues/6309

**Comment by DMartens** [comment]
This rule is already available in the core with [logical-assignment-operators](https://eslint.org/docs/latest/rules/logical-assignment-operators) without the type-awareness needed for `!foo`.

**Comment by OlivierZal** [comment]
@JoshuaKGoldberg, I'll take the challenge once the ongoing PR on this rule is merged 😊

**Comment by kirkwaiblinger** [comment]
> This rule is already available in the core with [logical-assignment-operators](https://eslint.org/docs/latest/rules/logical-assignment-operators) without the type-awareness needed for `!foo`.

That's an interesting point, actually. I was able to see this in the playground like so ([playground](https://typescript-eslint.io/play/#ts=5.7.2&fileType=.ts&code=CYUwxgNghgTiAEEQBd4DMD2GBc8De8UuAzsjAJYB2A5vAL7wA%2B8lArhBANwBQ3oksBGlaUwychkrwAtlADWIAGJYAFAEpcBIvFIUa9Ht2GjxkxFABeATwCSlcuKgRyFkACErAFRitkACyoQYmJ1fG54eHI0eBUAQkwMNTCIiIT4AF4ZeSVVNR4Ium5CoxExCSloazsHcicXdysAOXZnYj8AYT9wOVC8cMjolTT0zLYOJL6U9CwMrIVlDHV8%2BiLeYzKzStt7R2dXDwBZdnEAB1hkZo5yNs7u3v6omOGRlhamZmfMkVA0QOAJ-qpGaZWTzXLLQrFdamCqWbY1Or7KwAUQAHlAxJcILcwD0ARFHkNgS8xhB8SlhnMcos8v1IbwQKiThgYKgCHQgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WlqYSNkAC1pkA9gEMkyMswDm6KAHdJ0JpAA04bJHjj5lMtNqTkySvKYBbRC1rjS0SfnHRUGANo7sKtRs0wH11pVUJUbV8IYCCo3TsuNzJEADE3AEkuAGV8F0RbFg8wfDhEYIgAX2CAXR0qiqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)). The `if (!foo) foo = makeFoo()` case works indirectly by converting it to `foo ||= makeFoo()`, which prefer-nullish-coalescing then flags and suggests `foo ??= makeFoo()`. That seems fair enough.

One case that it can't get without type information is
```ts
declare let foo: { a: string } | null;

declare function makeFoo(): { a: string };

function lazyInitializeByExactNullCheck() {
  if (foo === null) {
    foo = makeFoo();
  }
}
```

It would be kind of nice to get that case too, so I think it makes sense to proceed with adding this check into prefer-nullish-coalescing. But maybe we should mention logical-assignment-operators in the docs?

**Comment by OlivierZal** [comment]
Should it report if the `ignoreConditionalTests` is enabled?

**Comment by OlivierZal** [comment]
@kirkwaiblinger, @JoshuaKGoldberg, I have a working branch: https://deploy-preview-10861--typescript-eslint.netlify.app/play/#ts=5.7.2&fileType=.tsx&code=CYUwxgNghgTiAEEQBd4DMD2GBc8De8UuAzsjAJYB2A5vAL7wA%2B8lArhBANwCwAUH6EiwEaVpTDJyGSvAC2UANYgAYlgAUASlwEi8UhRr0e-XqPGTpiKAC8AngElK5SVAjlrIAEK2AKjFbIABZUIMTEmvh88PDkaPBqAISYGBqRvNHRyfAAvHKKKuoaxtF0fKUmZhJSMtB2js7kru5etgBy7BAAwoHgChF4UTFxalnZuWwcqQPpGaN5SqoYmoPl5XwgAB4ADhgwqAR0QA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WlqYSNkAC1pkA9gEMkyMswDm6KL2jjokcGAC%2BILUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

I still need to do some refactoring and add tests + rebase once https://github.com/typescript-eslint/typescript-eslint/pull/10708 merged before to submit the PR.

**Comment by neilenns** [comment]
This should have gone out in a major release :( I woke up to failing CI builds and had to spend time finding out why a minor release of typescript-eslint started failing code that was previously fine.

**Comment by kirkwaiblinger** [comment]
@neilenns see conversation at https://github.com/typescript-eslint/typescript-eslint/issues/10995

**Comment by neilenns** [comment]
Yes, that's how I found this issue.
